src/register.rs

```diff
@@ -20,52 +20,65 @@ pub struct Register {
     /// The state vector is [a, b, c, ...]<sup>T</sup>, where |state_i|<sup>2</sup> represents the probability
     /// that the system will collapse into the state described by the ith basis vector.
     pub state: Array2<Complex<f64>>, // Should not be pub (it is pub now for testing purpouses)
-    size: usize
+    size: usize,
 }
 
-
 impl Register {
-
-    /// Creates a new state with an array of booleans with size N 
+    /// Creates a new state with an array of booleans with size N
     pub fn new(input_bits: &[bool]) -> Self {
         // Complex 1 by 1 identity matrix
-        let base_state = array![[Complex::new(1.0, 0.0)]]; 
-        
+        let base_state = array![[Complex::new(1.0, 0.0)]];
+
         // Creates state by translating bool to qubit
         // then uses qubits in tesnor product to create state
-        let state_matrix = input_bits.iter()
+        let state_matrix = input_bits
+            .iter()
             .map(math::to_qbit_vector)
             .fold(base_state, |a, b| linalg::kron(&b, &a));
 
         Self {
             state: state_matrix,
-            size: input_bits.len()
+            size: input_bits.len(),
         }
     }
     /// Applys a quantum operation to the current state
     ///
     /// Input a state and an operation. Outputs the new state
     pub fn apply(&mut self, op: &Operation) -> &mut Self {
-        // Gets the target bit
-        let target = op.targets()[op.arity() - 1];
-        // Calculates the number of matrices in tensor product
-        let num_matrices = self.size + 1 - op.targets().len();
+        // Permutation which the qubits will be shuffled according to
+        // Ex, perm[0]=3 means the qubit at idx 3 will be moved to idx 0
+        let mut perm = op.targets();
+        for i in 0..self.size {
+            if !perm.contains(&i) {
+                perm.push(i);
+            }
+        }
 
-        // If index i is equal to target bit returns matrix representation of operation
-        // otherwise returns 2 by 2 identity matrix
-        let get_matrix = |i| { if i == target { return op.matrix(); }
-            else { return operation::identity(0).matrix(); }
-        };
+        // Create a copy of state and permute the qubits according to perm
+        let mut permuted_state = self.state.clone();
+        for i in 0..permuted_state.len() {
+            let mut j: usize = 0;
+            for k in 0..self.size {
+                j |= ((i >> k) & 1) << perm[k];
```

Comment:
I think this code should be commented more. It is very hard to read and understand what it does other than that it permutes the qubits according to perm.